### PR TITLE
[CONTINT-3554] Send the `Datadog-Entity-ID` header, containing either the container-id or the cgroup inode if available

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR sends the `Datadog-Entity-ID` header when the `Datadog-Container-Id` is usually sent, containing either the container-id or the cgroup inode if available.

### Motivation
Using the cgroup inode will allow the trace-agent to retrieve the container-id, typically on cgroupv2 without UDS.

### Drawbacks
Datadog-Entity-ID will be sent even if the app is not running in a container because there is no reliable way to ensure that the app is running in a container.

### Test
This PR should be tested in several setups without UDS:
- Only cgroupv2 is available: Traces should contain the `in-<inode>` with the inode obtain by `kubectl exec <pod> -- stat /sys/fs/cgroup`
- Hybrid setup: If the container-id was available before this PR, it should still be available. Otherwise it should use the cgroup memory controller inode (`kubectl exec <pod> -- stat /sys/fs/cgroup/memory`).

With UDS, we should never use the inode logic.
- Cgroupv1: Same as hybrid

- To test e2e, use an image containing this PR https://github.com/DataDog/dd-trace-go/pull/2402 or directly `docker.io/alidatadog/agent:7-51-0-trace-containerid` on a kind 1.28 cluster using the following manifest:
```
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: dummy-apm-app
  spec:
    replicas: 1
    selector:
      matchLabels:
        app: dummy-apm-app
    template:
      metadata:
        labels:
          app: dummy-apm-app
          admission.datadoghq.com/enabled: "false"
      spec:
        containers:
        - name: dummy-apm-app
          image: docker.io/alidatadog/dummy-apm-app:0.2.0
          env:
          - name: DD_AGENT_HOST
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
```
- Otherwise I made a dummy app that logs span headers. It should be deployed on kind 1.28 as well. The docker vm should use cgroupv2 by default.
```
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: dummy-apm-app
  spec:
    replicas: 1
    selector:
      matchLabels:
        app: dummy-apm-app
    template:
      metadata:
        labels:
          app: dummy-apm-app
          admission.datadoghq.com/enabled: "false" # Important otherwise container_id won't be shared.
      spec:
        containers:
        - name: dummy-apm-app
          image: docker.io/alidatadog/dummy-apm-app:0.1.0
```
Example:
```
2023/11/30 11:10:50 map[Accept-Encoding:[gzip] Content-Type:[application/msgpack] Datadog-Client-Computed-Top-Level:[yes] Datadog-Client-Dropped-P0-Spans:[0] Datadog-Client-Dropped-P0-Traces:[0] Datadog-Entity-Id:[in-35385] Datadog-Meta-Lang:[go] Datadog-Meta-Lang-Interpreter:[gc-arm64-linux] Datadog-Meta-Lang-Version:[1.21.4] Datadog-Meta-Tracer-Version:[v1.59.0-dev] User-Agent:[Go-http-client/1.1] X-Datadog-Trace-Count:[1]]
```

- If you're interested by the app, I used the following code:
```
package main

import (
	"log"
	"net/http"
	"time"

	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func main() {
	go func() {
		http.HandleFunc("/v0.4/traces", func(w http.ResponseWriter, r *http.Request) {
			log.Println(r.Header)
		})
		log.Fatal(http.ListenAndServe(":8080", nil))
	}()

	tracer.Start(
		tracer.WithAgentAddr(":8080"),
		tracer.WithService("my.service"),
	)
	defer tracer.Stop()

	ticker := time.NewTicker(10 * time.Second)
	for ; true; <-ticker.C {
		span := tracer.StartSpan("my.span")
		time.Sleep(2 * time.Second)
		span.Finish()
	}
}
```
and build the image with the following `Dockerfile`

```
# Use the official Golang image to create a build artifact.
# This is based on Debian and sets the GOPATH to /go.
FROM golang:latest as builder

# Set the Current Working Directory inside the container
WORKDIR /app

# Copy the source from the current directory to the Working Directory inside the container
COPY . .

# Build the Go app
RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o main .

# Command to run the executable
CMD ["./main"]
``` 
To build a multi-arch image, I used `docker buildx build --push --platform=linux/amd64,linux/arm64 -t docker.io/alidatadog/dummy-apm-app:0.1.0 .`

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: internal/container.go
Comment (ahmed-mez) [inline internal/container.go:150]:
```suggestion
	cgroupMountPath := parseCgroupV2MountPath(f)
	if cgroupMountPath == "" {
		return ""
	}
	// Parse /proc/self/cgroup to retrieve the cgroup node path
	f, err = os.Open(cgroupPath)
	if err != nil {
		return ""
	}
	defer f.Close()
	cgroupNodePath := parseCgroupV2NodePath(f)
	if cgroupNodePath == "" {
		return ""
	}
	// Retrieve the cgroup inode from the cgroup mount and cgroup node path
	fi, err := os.Stat(path.Clean(cgroupMountPath + cgroupNodePath))
	if err != nil {
		return ""
	}
```
Style suggestion based on the repo guidelines https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines

**Review by ahmed-mez — COMMENTED** [review]
Looks great, thanks! Let's get this merged once we have the changes in the agent!

**Review by ahmed-mez — APPROVED** [review]

File: internal/container.go
Comment (ahmed-mez) [inline internal/container.go:147]: can we be extra safe here and do the type assertion properly?
```golang
    stat, ok := fi.Sys().(*syscall.Stat_t)
    if !ok {
        return ""
    }
    return fmt.Sprintf("in-%d", stat.Ino)
```

Comment (AliDatadog) [reply internal/container.go:147]: Sure, I don't think it could happen but it's better to be extra safe.

Comment (ahmed-mez) [reply internal/container.go]: Wondering if it would break on windows

Comment (AliDatadog) [reply internal/container.go]: It would fail much earlier I guess

**Review by piochelepiotr — APPROVED** [review]
LGTM for data streams

File: profiler/upload_test.go
Comment (nsrip-dd) [inline profiler/upload_test.go:49]: There's no assertion that the entity ID is empty. If I change this line to something non-empty the test still passes.

I'm not actually sure why this test asserts an empty container ID in the first place (which I assume you're following for consistency) but it doesn't _seem_ to be necessary. Do you think it's useful? If so let's comment why, otherwise I'd suggest dropping the change to this test.

Comment (AliDatadog) [reply profiler/upload_test.go:49]: Nice catch, I added the assertion to verify that the entityID is empty.

I am not familiar enough with this code base to know why it was there in the first place. Perhaps it is there to make sure that the functions called won't initialize the container-id or the entity-id. This way, if we refactor the code we will avoid introducing unexpected behaviors.

Comment (nsrip-dd) [reply profiler/upload_test.go:49]: 💡 Thought about it a little more, I think this is useful since it asserts that we won't have a header if there is no container ID or entity ID found, which seems to be the desired behavior. So, adding the assertion makes sense. Thanks for updating the test!

**Review by nsrip-dd — APPROVED** [review]
LGTM for profiling, thanks!

File: internal/container.go
Comment (vboulineau) [inline internal/container.go:156]: I think code is incorrect here, `mountsPath` is `/proc/mounts`, which is not read as we defaulted to always reading `/sys/fs/cgroup`, it should be `defaultCgroupMountPath` and `mountsPath` should be removed.

Note that you have no unit test on that.

Comment (AliDatadog) [reply internal/container.go]: Nice catch. I fixed it and added a unit test

**Comment by ahmed-mez** [comment]
@AliDatadog @vboulineau is this ready to merge?

**Comment by AliDatadog** [comment]
> @AliDatadog @vboulineau is this ready to merge?

Waiting for the last approval

File: internal/container.go
Comment (vboulineau) [inline internal/container.go:24]: Not used, should be removed

**Review by vboulineau — APPROVED** [review]

**Review by vboulineau — APPROVED** [review]

**Pull request merged**
